homeassistant/components/lifx/light.py

```diff
@@ -378,18 +379,21 @@ async def start_effect(self, entities, service, **kwargs):
 
     @callback
     def register(self, bulb):
-        """Handle aiolifx detected bulb."""
-        self.hass.async_create_task(self.register_new_bulb(bulb))
-
-    async def register_new_bulb(self, bulb):
-        """Handle newly detected bulb."""
+        """Allow a single in-flight discovery per bulb."""
+        if bulb.mac_addr not in self.discoveries_inflight:
```

Comment:
I have a few theories on this as I see the same thing happening with my Photons-powered integration. I _think_ it's because it relies on the zeroconf/HomeKit mDNS to trigger discovery instead of handling it internally, and it may be getting triggered multiple times.

Reply:
My gut feeling is that this is the real issue and this PR is papering over it so it won't get properly fixed.

From your "home-assistant-current.log":
```
2022-04-25 09:45:52 DEBUG (MainThread) [homeassistant.components.lifx.light] 192.168.42.203 register NEW
[...]
2022-04-25 09:45:53 DEBUG (MainThread) [homeassistant.components.lifx.light] 192.168.42.203 register NEW
```

The above should never happen since discoveries are 1 minute apart. Also, the race in `register_new_bulb()` should not be an issue because aiolifx guards us.

All your bulbs register twice so I suspect that you are somehow running two discovery tasks. Unfortunately, I do not see that issue myself and the logging is very sparse so I am unable to figure out what happens :-(

(I can reproduce the "unique IDs" issue by setting an IP alias on my network interface so ifaddr finds two addresses with the same broadcast address)

Maybe this could help to uncover the underlying issue?

```diff
--- a/aiolifx/aiolifx.py
+++ b/aiolifx/aiolifx.py
@@ -1431,6 +1431,9 @@ class LifxScan:
             lifx_discovery = LifxDiscovery(self.loop, manager)
             discoveries.append(lifx_discovery)
             lifx_discovery.start(listen_ip=ip)
+            print(f"Starting discovery on {ip}")
+            import traceback
+            traceback.print_stack()
             tasks.append(self.loop.create_task(manager.lifx_ip()))

         (done, pending) = await aio.wait(tasks, timeout=timeout)
```

Reply:
Yes, I am running multiple discovery tasks because I have both a wired and wifi connection on my development machine. I suspect this happens a lot though, as many Raspberry Pis would be configured the same way.

In my Photons-based integration, I determine the preferred interface for each broadcast network, but `aiolifx` doesn't do this. It just happily starts multiple scans. By limiting the discoveries in-flight, this eliminates the race condition that occurs on misconfigured multi-homed devices.

Reply:
I don't see this as papering over the real issue, though. I can't change how people chose to configure their systems (and it seems I have to take another look at how macOS is configured wrt networking), but this refactor at least nullifies its impact.

Reply:
The setup process with two discoveries is:
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: discovery request -> 60x discovery response -> feature request -> feature response -> [duplicate ID]
```

This PR proposes to ignore the second discovery response. That saves the unicast feature request for the duplicate entry (and thus avoids the duplicate ID warning):
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: discovery request -> 60x discovery response -> [ignore]
```

However, what we really want is just a single discovery since that will furthermore save a redundant broadcast and all the matching responses from each bulb:
```
  wired: discovery request -> 60x discovery response -> feature request -> feature response -> [add entity]
  wifi: [idle]
```

So I think the bug is with the network detection (i.e. the ifaddr usage in aiolifx), not really `register_new_bulb()`.

Reply:
Fair enough. I think I can leverage HA's own networking helpers to limit discovery to whichever interface has been selected for mDNS by HA, and then limit aiolifx to that network interface.

Reply:
Awesome. You might find that adjusting the retries does not matter then because bulbs do not get overloaded in the first place.

Reply:
@amelchio four years ago, you added this line:
```
lifx_ip_addresses = await aiolifx().LifxScan(hass.loop).scan()
```

Any objection if I just remove this line and instead, only enable LIFX discovery on network interfaces enabled by Home Assistant?

Reply:
That sounds reasonable. HA did not have any interface selector back then :)

It's a breaking change, of course – but not in a bad way.

Reply:
How much do we want to protect users from themselves? Specifically: if someone choses to manually enable multiple interfaces on the same broadcast network, the current behaviour will re-occur, unless we specifically check for that as well. 

I'm actually happy with this option because I'd document the badness that would occur if such a configuration is chosen by the user and it would mean we could just let Home Assistant's network config be authorative. 

Thoughts?

Reply:
Alternatively, I keep this fix in place as well, which means if someone _does_ enable multiple interfaces that trigger duplicate discovery, we at least ignore some of it at the cost of additional network traffic.

Reply:
For that situation I would propose my one-line fix from yesterday but upgraded to a warning since we now realized that it is a situation to avoid: https://github.com/home-assistant/core/pull/70458#discussion_r857101697

I personally think that's a fine balance between usability and code complexity. However, @bdraco can probably comment on how other integrations handle this (i.e. being enabled for wired/wifi interfaces on the same subnet)

Reply:
I would personally not complicate the code with the in-flight dictionary just to cater for misconfigurations – but I think that is the kind of judgement call that a code owner should be making :-)

Reply:
I see what you did there. :)

Reply:
Checking the configured network interfaces is a nice improvement 👍 

You'll definitely find that you'll still have the same issue since there are still all kinds of "strange" network setups.  We end up with lots of code in zeroconf to handle these type of problems with overlapping responses since even with the ability to configure it, its still quite a common issue.

As much as it would be nice to to have these type of issues, I think you are stuck with either tracking the discoveries to avoid dupes as you have now, or some type of locking (usually less preferred).   I've tried to remove similar code in zeroconf and always end up having to put it back.

Here is a section from RFC6762 that highlights how different implementations can be https://www.rfc-editor.org/rfc/rfc6762.html#section-6.2
```
   Some Multicast DNS responders treat a physical interface with both
   IPv4 and IPv6 address as a single interface with two addresses.
   Other Multicast DNS responders may treat this case as logically two
   interfaces (one with one or more IPv4 addresses, and the other with
   one or more IPv6 addresses), but responders that operate this way
   MUST NOT put the corresponding automatic NSEC records in replies they
   send (i.e., a negative IPv4 assertion in their IPv6 responses, and a
   negative IPv6 assertion in their IPv4 responses) because this would
   cause incorrect operation in responders on the network that work the
   former way.
```

Reply:
Yeah, it makes quite the difference for performance and stability, too. I'm happy keeping the in-flight check as a backup for those network configs that are on the more insane side.

Reply:
@bdraco Can you clarify a bit ... I understand your experience that a request can have multiple responses but is the change here (i.e. to use the configured network interfaces) sufficient to keep _us_ from sending multiple requests in the first place?

Reply:
I am not asking whether people will shoot their feet but rather if they are able to avoid it.

(Like I said, I'm down with you change; just trying to fully understand the situation)

Reply:
Gotcha. It's certainly avoidable: just leave the network config on Auto, or ensure that if you do have multiple interfaces enabled, they're on different IP networks.

Reply:
It is avoidable except when they actually need to broadcast on multiple interfaces because they have multiple subnets.  We have users that have their IoT subnet, their secure subnet, etc so we should expect they are going to turn on multiple interfaces and we may get broadcast traffic from one reflected back on another.

Reply:
Multiple distinct subnets should be fine as LIFX uses UDP broadcast for discovery, which is rarely reflected, repeated or routed.

Multiple enabled interfaces on the same subnet however, will generate duplicate responses.

This is usually what happens when someone enables both wifi and a wired connection on an RPi, for example.

Reply:
> Multiple distinct subnets should be fine as LIFX uses UDP broadcast for discovery, which is rarely reflected, repeated or routed.

Take a look at the python-zeroconf issue queue history, it paints a slightly different view.

> 
> Multiple enabled interfaces on the same subnet however, will generate duplicate responses.

Yes

> 
> This is usually what happens when someone enables both wifi and a wired connection on an RPi, for example.

If you have wired I would considering enabling wifi as well a bad idea but it happens all the time unfortunately.